File: spark-scala/spark-dataframe/dataframe-conversion/src/main/scala/com/mahfooz/dataframe/rdd/DataframeFromRdd.scala
package com.mahfooz.dataframe.rdd

import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.sql.types.{
  LongType,
  StringType,
  StructField,
  StructType
}

object DataframeFromRdd {

  def main(args: Array[String]): Unit = {
    val spark = SparkSession
      .builder()
      .master("local[*]")
      .appName("DataframeFromRdd")
      .getOrCreate()

    val schema = StructType(
      Array(
        StructField("id", LongType, nullable = true),
        StructField("name", StringType, nullable = true),
        StructField("age", LongType, nullable = true)
      )
    )

    val peopleRDD = spark.sparkContext.parallelize(
      Array(Row(1L, "John Doe", 30L), Row(2L, "Mary Jane", 25L))
    )

    val peopleDF = spark.createDataFrame(peopleRDD, schema)

    peopleDF.printSchema

    peopleDF.show()
  }

}

File: spark-java/spark-dataframe/dataframe-conversion/src/main/java/com/mahfooz/df/rdd/DataframeFromRdd.java
package com.mahfooz.df.rdd;

import com.mahfooz.df.model.Person;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

public class DataframeFromRdd {

    public static void main(String[] args) {

        SparkSession spark = SparkSession
                .builder()
                .master("local[*]")
                .appName("Dataframe From RDD")
                .getOrCreate();
        String data_file = System.getenv("DATA_HOME")+"/FileData/Text/people.txt";

        // Create an RDD of Person objects from a text file
        JavaRDD<Person> peopleRDD = spark.read()
                .textFile(data_file)
                .javaRDD()
                .map(line -> {
                    String[] parts = line.split(",");
                    Person person = new Person();
                    person.setId(Integer.parseInt(parts[0].trim()));
                    person.setName(parts[1]);
                    person.setAge(Integer.parseInt(parts[2].trim()));
                    return person;
                });

        // Apply a schema to an RDD of JavaBeans to get a DataFrame
        Dataset<Row> peopleDF = spark.createDataFrame(peopleRDD, Person.class);
        peopleDF.explain();
        peopleDF.show();
    }
}
